File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    List<Integer> list = new ArrayList<Integer>();
    public boolean isValidBST(TreeNode root) {
        inOrder(root);
        int temp = Integer.MIN_VALUE;
        for(Integer i: list){
            if(i <= temp){
                return false;
            }
            temp = i;
        }

        return true;
    }

    private void inOrder(TreeNode root){
        if(root ==null){
            return;
        }
        //left
        inOrder(root.left);
        //root
        list.add(root.val);
        //right
        inOrder(root.right);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    var ans = 0;
    fun diameterOfBinaryTree(root: TreeNode?): Int {    
        if(root == null){
            return 0;
        }

        diameterHelper(root)
        return ans
    }

    fun diameterHelper(root: TreeNode?){

        if(root == null){
            return
        }
        
        diameterHelper(root.left)
        diameterHelper(root.right)


        val leftDepth = maxDepthMeasure(root.left)
        val rightDepth =  maxDepthMeasure(root.right);

        ans = max(ans, leftDepth+rightDepth);

    }

    fun maxDepthMeasure(root: TreeNode?): Int{
        if(root == null){
            return 0
        }

        return max(maxDepthMeasure(root.left), maxDepthMeasure(root.right))+1;
    }
}
